runtime/near-wallet-contract/implementation/wallet-contract/src/internal.rs

```diff
@@ -0,0 +1,286 @@
+                public_key,
+            })
+        }
+        _ => eth_emulation::try_emulation(target, tx, context),
+    }
+}
+
+/// Validates the transaction is following the Wallet Contract protocol.
+/// This includes checks for:
+/// - from address matches current account address
+/// - to address is present and matches the target address (or hash of target account ID)
+/// - nonce matches expected nonce
+/// If this validation fails then the relayer that sent it is faulty and should be banned.
+fn validate_tx_relayer_data(
+    tx: &NormalizedEthTransaction,
+    target: &AccountId,
+    context: &ExecutionContext,
+    expected_nonce: u64,
+) -> Result<TransactionValidationOutcome, Error> {
+    if tx.address.raw() != context.current_address {
+        return Err(Error::Relayer(RelayerError::InvalidSender));
+    }
+
+    if tx.chain_id != Some(CHAIN_ID) {
+        return Err(Error::Relayer(RelayerError::InvalidChainId));
+    }
+
+    let to = tx
+        .to
+        .ok_or(Error::User(UserError::EvmDeployDisallowed))?
+        .raw();
+    let target_as_address = extract_address(target).ok();
+    let to_equals_target = target_as_address
+        .map(|target| to == target)
+        .unwrap_or(false);
+
+    // Only valid targets satisfy `to == target` or `to == hash(target)`
+    if !to_equals_target && to != hash_to_address(target) {
+        return Err(Error::Relayer(RelayerError::InvalidTarget));
+    }
+
+    let nonce = if tx.nonce <= U64_MAX {
+        tx.nonce.low_u64()
+    } else {
+        return Err(Error::Relayer(RelayerError::InvalidNonce));
+    };
+    if nonce != expected_nonce {
+        return Err(Error::Relayer(RelayerError::InvalidNonce));
+    }
+
+    // If `to == target` and this is not a self-transaction then the address must not
+    // be registered in the address registry. The purpose of this check is to prevent
+    // lazy relayers from skipping this check themselves (relayers are supposed to use
+    // the address registry to fill in the `target`).
+    if to_equals_target && to != context.current_address {
+        Ok(TransactionValidationOutcome::AddressCheckRequired(to))
+    } else {
+        Ok(TransactionValidationOutcome::Validated)
+    }
+}
+
+fn validate_tx_value(
+    tx: &NormalizedEthTransaction,
+    context: &ExecutionContext,
+    action: &Action,
+) -> Result<(), Error> {
+    if tx.value.raw() > VALUE_MAX {
+        return Err(Error::User(UserError::ValueTooLarge));
+    }
+
+    let total_value = tx
+        .value
+        .raw()
+        .low_u128()
+        .saturating_mul(MAX_YOCTO_NEAR.into())
+        .saturating_add(action.value().as_yoctonear());
+
+    if total_value > 0 {
+        let is_self_call = context.predecessor_account_id == context.current_account_id;
+        let sufficient_attached_deposit =
+            context.attached_deposit >= NearToken::from_yoctonear(total_value);
+        if !is_self_call && !sufficient_attached_deposit {
```

Comment:
No matter what, it is always true that any action is signed by the owner of the address because `rlp_execute` starts by decoding the RLP-encoded Ethereum transaction and checking its signature. The purpose of this check is to force external relayers to pay for the Near all fees associated with a transaction, not just the gas. This is similar to how meta-transactions work for normal (not eth-implicit) Near accounts.

The reason to allow self-calls is to provide a pathway for relayers to make users pay for their own Near fees. The flow would be for the relayer to have the user sign a transaction adding a function call access key which the relayer controls to the eth-implicit account. Then the relayer can sign transactions with that access key to have the wallet contract call its own `rlp_execute` method (of course the payload to that method must still be an Ethereum transaction signed by the user's key; so the relayer cannot use this to make the account do something the user did not intend).